### PR TITLE
Accordion have the ability to expand certain card

### PR DESCRIPTION
**Pull request by hoaaah**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | no
| Fixed issues  | #158 

This PR fixes some issue in accordion. Recreate widget based on  [https://getbootstrap.com/docs/4.0/components/collapse/#accordion-example].

Give ability to set expanded item, so it change behavior from this

```php
'contentOptions' => ['class' => 'show']
```
to this

```php
'expand' => true, // default to false
```

### CODE REVIEW

File: src/Accordion.php [hunk 1]
```diff
@@ -170,9 +184,10 @@ public function renderItem($header, $item, $index)
             $options = ArrayHelper::getValue($item, 'contentOptions', []);
             $options['id'] = $id;
             Html::addCssClass($options, ['widget' => 'collapse']);
-            if ($index === 0) {
```
Comment (Thoulah) [inline src/Accordion.php:173]: Please watch line 190/205 as well (`aria-expanded`)

Comment (hoaaah) [reply src/Accordion.php:173]: is it like this? Not really sure about this.
```php
'aria-expanded' => ($index === 0) ? 'true' : 'false',
```

File: src/Accordion.php [hunk 2]
```diff
@@ -170,9 +184,10 @@ public function renderItem($header, $item, $index)
...
-                Html::addCssClass($options, 'show');
-            }
+
+            // check if accordion expanded, if true add show class
+            if($this->_isAccordionExpanded($item)) $options['class'] = ' show';
```
Comment (Thoulah) [inline src/Accordion.php:189]: Please use full syntax here, yii2 supports PHP 5.4 and higher.

File: src/Accordion.php [hunk 3]
```diff
@@ -154,6 +155,19 @@ public function renderItems()
...
     }
 
+    /**
+     * check wether accordion is expanded
+     * This method will check if expand properties exist and determine value of this properties
+     * 
+     * @return bool
+     */
+    private function _isAccordionExpanded($item){
+        if(isset($item['expand'])){
```
Comment (Thoulah) [inline src/Accordion.php:165]: `if (isset($item['expand'])) {`

File: src/Accordion.php (hunk 1 above)
Comment (Thoulah) [reply src/Accordion.php:173]:
```php
'aria-expanded' => ($this->_isAccordionExpanded($item)) ? 'true' : 'false',
```

File: src/Accordion.php (hunk 3 above)
Comment (hoaaah) [reply src/Accordion.php:166]: Fix in this https://github.com/yiisoft/yii2-bootstrap4/pull/196/commits/65d25c75c39e69766b38f2fdd1e292b72325c546

**Comment by Thoulah** [comment]
If I'm correct, the first card is now never expanded/shown if no expand is given. This is a change from the default behaviour (hence the failing tests).
Please also add tests with this new behaviour.

File: src/Accordion.php [hunk 4]
```diff
@@ -154,6 +155,20 @@ public function renderItems()
...
     }
 
+    /**
+     * check wether accordion is expanded
+     * This method will check if expand properties exist and determine value of this properties
+     * 
+     * @param array $item
+     * @return bool
+     */
+    private function _isAccordionExpanded($item){
```
Comment (simialbi) [inline src/Accordion.php:165]: Methods (even private ones) aren't prefixed with underline in [Yii2 code styles](https://github.com/yiisoft/yii2/blob/master/docs/internals/core-code-style.md#43-methods) and the brace of a method should be placed on the next line.

File: src/Accordion.php [hunk 5]
```diff
@@ -154,6 +155,20 @@ public function renderItems()
...
+        if(isset($item['expand'])){
```
Comment (simialbi) [inline src/Accordion.php:166]: There should be spaces after `if` statement and between `)` and `{` according [Yii code styles](https://github.com/yiisoft/yii2/blob/master/docs/internals/core-code-style.md#54-control-statements).

File: src/Accordion.php [hunk 6]
```diff
@@ -170,9 +185,12 @@ public function renderItem($header, $item, $index)
             $options = ArrayHelper::getValue($item, 'contentOptions', []);
             $options['id'] = $id;
             Html::addCssClass($options, ['widget' => 'collapse']);
-            if ($index === 0) {
-                Html::addCssClass($options, 'show');
+
+            // check if accordion expanded, if true add show class
+            if($this->_isAccordionExpanded($item)){
```
Comment (simialbi) [inline src/Accordion.php:190]: There should be spaces after `if` statement and between `)` and `{` according [Yii code styles](https://github.com/yiisoft/yii2/blob/master/docs/internals/core-code-style.md#54-control-statements).

File: src/Accordion.php [hunk 7]
```diff
@@ -170,9 +185,12 @@ public function renderItem($header, $item, $index)
...
+                $options['class'] = ' show';
```
Comment (simialbi) [inline src/Accordion.php:191]: Like this you override existing classes like [`collapse` class](https://github.com/yiisoft/yii2-bootstrap4/pull/196/files#diff-0450c68805ce55d68227126fef5003bbR187). Replace with `Html::addCssClass($options, 'show');` will preserve them.

**Comment by hoaaah** [comment]
> 
> 
> If I'm correct, the first card is now never expanded/shown if no expand is given. This is a change from the default behaviour (hence the failing tests).
> Please also add tests with this new option.

I don't think I have enough competence to modify this test. Please review this commit https://github.com/yiisoft/yii2-bootstrap4/pull/196/commits/d1233fc1cf5799af78520894f4fd3130931ffa6d and if anyone have proper test, please give your suggestion.

**Comment by Thoulah** [comment]
You could have just accepted [my pull request](https://github.com/BosSTAN-id/yii2-bootstrap4/pull/1) on your repo.

**Comment by hoaaah** [comment]
> 
> 
> You could have just accepted [my pull request](https://github.com/BosSTAN-id/yii2-bootstrap4/pull/1) on your repo.

Merged!. Thank you very much.

File: tests/AccordionTest.php [hunk 8]
```diff
@@ -165,6 +166,44 @@ public function testLabelKeys()
...
 </div>
 
+HTML
+        , $output);
+    }
+
+    public function testExpandOptions()
+    {
+        ob_start();
+        $form = ActiveForm::begin(['action' => '/something']);
```
Comment (Thoulah) [inline tests/AccordionTest.php:176]: My mistake, but delete this :)

File: tests/AccordionTest.php [hunk 9]
```diff
@@ -165,6 +166,44 @@ public function testLabelKeys()
...
+        ActiveForm::end();
```
Comment (Thoulah) [inline tests/AccordionTest.php:177]: My mistake, but delete this :)

File: tests/AccordionTest.php [hunk 10]
```diff
@@ -165,6 +166,44 @@ public function testLabelKeys()
...
+        ob_end_clean();
```
Comment (Thoulah) [inline tests/AccordionTest.php:178]: My mistake, but delete this :)

File: tests/AccordionTest.php [hunk 11]
```diff
@@ -165,6 +166,44 @@ public function testLabelKeys()
...
+
```
Comment (Thoulah) [inline tests/AccordionTest.php:179]: My mistake, but delete this :)

File: tests/AccordionTest.php [hunk 12]
```diff
@@ -73,7 +74,7 @@ public function testRender()
 <div id="w0" class="accordion">
 <div class="card"><div id="w0-collapse0-heading" class="card-header"><h5 class="mb-0"><button type="button" id="w1" class="btn-link btn" data-toggle="collapse" data-target="#w0-collapse0" aria-expanded="true" aria-controls="w0-collapse0">Collapsible Group Item #1</button>
 </h5></div>
-<div id="w0-collapse0" class="collapse show" aria-labelledby="w0-collapse0-heading" data-parent="#w0">
+<div id="w0-collapse0" class="collapse" aria-labelledby="w0-collapse0-heading" data-parent="#w0">
```
Comment (Thoulah) [inline tests/AccordionTest.php:77]:
```suggestion
<div id="w0-collapse0" class="collapse show" aria-labelledby="w0-collapse0-heading" data-parent="#w0">
```

File: tests/AccordionTest.php [hunk 13]
```diff
@@ -32,6 +32,7 @@ public function testRender()
                         'class' => 'testClass',
                         'id' => 'testId'
                     ],
+                    'expand' => true,
```
Comment (Thoulah) [inline tests/AccordionTest.php:35]:
```suggestion
```

File: tests/AccordionTest.php [hunk 14]
```diff
@@ -82,7 +83,7 @@ public function testRender()
 </div></div>
 <div id="testId" class="testClass card"><div id="w0-collapse1-heading" class="card-header"><h5 class="mb-0"><button type="button" id="w2" class="btn-link btn" data-toggle="collapse" data-target="#w0-collapse1" aria-expanded="false" aria-controls="w0-collapse1">Collapsible Group Item #2</button>
 </h5></div>
-<div id="w0-collapse1" class="testContentOptions collapse" aria-labelledby="w0-collapse1-heading" data-parent="#w0">
+<div id="w0-collapse1" class="testContentOptions collapse show" aria-labelledby="w0-collapse1-heading" data-parent="#w0">
```
Comment (Thoulah) [inline tests/AccordionTest.php:86]:
```suggestion
<div id="w0-collapse1" class="testContentOptions collapse" aria-labelledby="w0-collapse1-heading" data-parent="#w0">
```

File: tests/AccordionTest.php [hunk 15]
```diff
@@ -165,6 +166,44 @@ public function testLabelKeys()
 </div></div>
 </div>
 
+HTML
+        , $output);
+    }
+
+    public function testExpandOptions()
+    {
+        ob_start();
```
Comment (Thoulah) [inline tests/AccordionTest.php:175]: My mistake, but delete this :)

**Comment by Thoulah** [comment]
I don't like my way of finding `$hasExpand` [here](https://github.com/yiisoft/yii2-bootstrap4/pull/196/files#diff-0450c68805ce55d68227126fef5003bbR141). Anyone has a better method?

**Comment by simialbi** [comment]
> I don't like my way of finding `$hasExpand` [here](https://github.com/yiisoft/yii2-bootstrap4/pull/196/files#diff-0450c68805ce55d68227126fef5003bbR141). Anyone has a better method?

I tried to simplify it a little bit:
https://github.com/BosSTAN-id/yii2-bootstrap4/pull/2

@hoaaah I opened a pull request in your fork

**Comment by simialbi** [comment]
I produced a BC break with the previous pull request so I had to fix it with another one

**Comment by hoaaah** [comment]
Is there any updates on this PR will be merged?

**Comment by samdark** [comment]
There are conflicts to be resolved. After it is done and if @simialbi and @Thoulah are fine with the it, I'll hit the merge button.

**Comment by simialbi** [comment]
@hoaaah I opened a pull request to do this 4 you: https://github.com/BosSTAN-id/yii2-bootstrap4/pull/4

File: src/Accordion.php [hunk 16]
```diff
@@ -159,20 +163,24 @@ public function renderItems()
      * @param string $header a label of the item group [[items]]
      * @param array $item a single item from [[items]]
      * @param int $index the item index as each item group content must have an id
+     * @param bool $expand whether or not to expand the item
      * @return string the rendering result
      * @throws InvalidConfigException
      * @throws \Exception
      */
-    public function renderItem($header, $item, $index)
+    public function renderItem($header, $item, $index, $expand)
```
Comment (xepozz) [inline src/Accordion.php:171]: Can you also typify the parameters?

Comment (Thoulah) [reply src/Accordion.php:171]: Based on what [guideline](https://github.com/yiisoft/yii2/blob/master/docs/internals/core-code-style.md)?

Comment (samdark) [reply src/Accordion.php:171]: @xepozz that's Yii 2. We're not using types there.

**Comment by Thoulah** [comment]
> There are conflicts to be resolved. After it is done and if @simialbi and @Thoulah are fine with the it, I'll hit the merge button.

Code looks awesome, tests are passing. I'm go

**Comment by simialbi** [comment]
I agree with @Thoulah

File: src/Accordion.php (hunk 16 above)
Comment (samdark) [reply src/Accordion.php:171]: That's change of signature of public method. That could potentially result in:

1. Breaks when the method is called w/o passing additional parameter. That could be fixed by introducing default value such as `$expand = false`.
2. Breaks if there's class extending `Accordion`. That would result in "Fatal error: Declaration of Accordion::renderItem($header, $item, $index, $expand) must be compatible with Accordion::renderItem($header, $item, $index, $expand)".

If the introduction of this parameter cannot be avoided, the change can be released in 2.1.0 only.

**Comment by samdark** [comment]
See https://github.com/yiisoft/yii2-bootstrap4/pull/196#discussion_r455346655

File: src/Accordion.php (hunk 16 above)
Comment (Thoulah) [reply src/Accordion.php:171]: @simialbi `$this->expand` perhaps?

Comment (simialbi) [reply src/Accordion.php:171]: I would prefer `$item['expand']`
```suggestion
    public function renderItem($header, $item, $index)
    {
         if (array_key_exists('content', $item)) {
             $id = $this->options['id'] . '-collapse' . $index;
             $options = ArrayHelper::getValue($item, 'contentOptions', []);
             $expand = ArrayHelper::getValue($item, 'expand', false); // <= Like this
```

Comment (simialbi) [reply src/Accordion.php:171]: But with a deeper look it seems absurd. Get the item with the expand property to set it a bit further again. Maybe we should remove this check of expanded item:
```php
    $expanded = array_search(true, ArrayHelper::getColumn(ArrayHelper::toArray($this->items), 'expand', true));
    if ($expanded === false) {
        $expanded = current(array_keys($this->items));
    }
```
And replace it with some simpler check if there is a `expanded => true` item and if it's not, set the first one (for **BC**). That would even eliminate the need for setting the `expand` property in the loop (from my suggestion above).

**Comment by simialbi** [comment]
I changed the code to prevent the method signature change:
https://github.com/BosSTAN-id/yii2-bootstrap4/pull/5
I tried different methods of resolution and finished with this one. I'm not really happy, but it works...
@Thoulah, @samdark: What do you think?

PS: this solution does not prevent multiple expand items any more, but I think it's possible to delegate this responsability to the developer. It's not even a BC break.

**Comment by Thoulah** [comment]
I think it would be __nice__ to have a max of 1 expansion, but indeed not required: devs _should_ take care of this themselves.
Besides that: good solution!

**Comment by samdark** [comment]
@simialbi looks way better about backwards compatibility. What was the test removed?

**Comment by Thoulah** [comment]
> @simialbi looks way better about backwards compatibility. What was the test removed?

Only `testExpandOnlyOneItem` was removed, as mentioned in his PS [here](https://github.com/yiisoft/yii2-bootstrap4/pull/196#issuecomment-659221510). `testExpandOptions` is still in place.

**Comment by simialbi** [comment]
@hoaaah Could you merge [my pull request](https://github.com/BosSTAN-id/yii2-bootstrap4/pull/5) in your repository please?

**Comment by hoaaah** [comment]
> 
> 
> @hoaaah Could you merge [my pull request](https://github.com/BosSTAN-id/yii2-bootstrap4/pull/5) in your repository please?

Merged! Really great solution!

**Comment by samdark** [comment]
Thanks!

**Pull request merged**